src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl_ModelComputation.cs

```diff
@@ -0,0 +1,256 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    internal partial class DocumentOutlineControl
+    {
+        private IWpfTextView? GetLastActiveIWpfTextView()
+        {
+            ThreadingContext.ThrowIfNotOnUIThread();
+
+            if (ErrorHandler.Failed(CodeWindow.GetLastActiveView(out var textView)))
+                return null;
+
+            return EditorAdaptersFactoryService.GetWpfTextView(textView);
+        }
+
+        /// <summary>
+        /// Starts a new task to compute the data model.
+        /// </summary>
+        private void StartComputeDataModelTask()
+        {
+            // 'true' value is unused.  this just signals to the queue that we have work to do.
+            _computeDataModelQueue.AddWork(true);
+        }
+
+        /// <summary>
+        /// Creates the data model.
+        /// </summary>
+        private async ValueTask<DocumentSymbolModel?> ComputeDataModelAsync(ImmutableSegmentedList<bool> _, CancellationToken cancellationToken)
+        {
+            // Jump to the UI thread to get the currently active text view.
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var activeTextView = GetLastActiveIWpfTextView();
+            if (activeTextView is null)
+                return null;
+
+            var originalSnapshot = activeTextView.TextSnapshot;
+            var textBuffer = activeTextView.TextBuffer;
+
+            var filePath = GetFilePath();
+            if (filePath is null)
+                return null;
+
+            // Ensure we switch to the threadpool before calling ComputeUIModelAsync. It ensures
+            // that fetching and processing the document model is not done on the UI thread.
+            await TaskScheduler.Default;
+
+            var model = await ComputeDataModelAsync().ConfigureAwait(false);
```

Comment:
I admit this threw me a bit until I saw this was a local method -- was surprised this didn't have a cancellation token going to it. Do we need this to be a local method or can we just inline it?

Reply:
I made sure to add `cancellationToken.ThrowIfCancellationRequested()` at the top of the local method and renamed it for clarity but I don't want to inline it since the same thing is done in `NavigationBarController_ModelComputation.cs`